src/Widgets/ScreenShield.vala

```diff
@@ -0,0 +1,433 @@
+                expand_to_screen_size ();
+            }
+        }
+
+        // status becomes idle after interval defined at /org/gnome/desktop/session/idle-delay
+        private void on_status_changed (PresenceStatus status) {
+            if (status != PresenceStatus.IDLE || !connected_to_buses) {
+                return;
+            }
+
+            debug ("session became idle, activating screensaver");
+
+            activate (true);
+        }
+
+        // We briefly inhibit sleep so that we can try and lock before sleep occurs if necessary
+        private void sync_inhibitor () {
+            if (!connected_to_buses) {
+                return;
+            }
+
+            var lock_enabled = gala_settings.get_boolean (LOCK_ON_SUSPEND_KEY);
+            var lock_prohibited = lockdown_settings.get_boolean (LOCK_PROHIBITED_KEY);
+
+            var inhibit = login_session != null && login_session.active && !active && lock_enabled && !lock_prohibited;
+            if (inhibit) {
+                try {
+                    var new_inhibitor = login_manager.inhibit ("sleep", "Pantheon", "Pantheon needs to lock the screen", "delay");
+                    if (inhibitor != null) {
+                        inhibitor.close ();
+                        inhibitor = null;
+                    }
+
+                    inhibitor = new_inhibitor;
+                } catch (Error e) {
+                    warning ("Unable to inhibit sleep, may be unable to lock before sleep starts: %s", e.message);
+                }
+            } else {
+                if (inhibitor != null) {
+                    try {
+                        inhibitor.close ();
+                    } catch (Error e) {
+                        warning ("Unable to remove sleep inhibitor: %s", e.message);
+                    }
+
+                    inhibitor = null;
+                }
+            }
+        }
+
+        private void on_user_became_active () {
+            if (!connected_to_buses) {
+                return;
+            }
+
+            // User became active in some way, switch to the greeter if we're not there already
+            if (is_locked && !in_greeter) {
+                debug ("user became active, switching to greeter");
+                cancel_animation ();
+                try {
+                    display_manager.switch_to_greeter ();
+                } catch (Error e) {
+                    critical ("Unable to switch to greeter to unlock: %s", e.message);
+                }
+                in_greeter = true;
+            // Otherwise, we're in screensaver mode, just deactivate
+            } else if (!is_locked) {
+                debug ("user became active in unlocked session, closing screensaver");
+                deactivate (false);
+            }
+        }
+
+        private LoginSessionManager? get_current_session_manager () throws GLib.Error {
+            string? session_id = GLib.Environment.get_variable ("XDG_SESSION_ID");
```

Comment:
```suggestion
            unowned string? session_id = GLib.Environment.get_variable ("XDG_SESSION_ID");
```
Unsure if that's possible with `login_user_manager.display.session` though

Reply:
Yeah, unfortunately this doesn't work with `login_user_manager.display.session`